### PR TITLE
[BPK-1383] Reduce duplication of theming logic

### PR DESCRIPTION
**Pull request by k0nserv**

### CODE REVIEW

File: native/packages/react-native-bpk-component-button-link/src/BpkButtonLink.android.js [hunk 1]
```diff
@@ -52,10 +52,11 @@ const BpkButtonLink = (props: Props) => {
...
 
-  const themeStyle =
-    theme && themeAttributesSupplied(theme)
-      ? { color: theme.buttonLinkTextColor }
-      : null;
+  const themeAttributes = getThemeAttributes(REQUIRED_THEME_ATTRIBUTES, theme);
+  let themeStyle = null;
+  if (themeAttributes) {
+    themeStyle = { color: themeAttributes.buttonLinkTextColor };
+  }
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-button-link/src/BpkButtonLink.android.js]: Why not:
```js
const themeStyle = themeAttributes ? { color: themeAttributes.buttonLinkTextColor } : null;
```
?

File: native/packages/react-native-bpk-component-button-link/src/BpkButtonLink.ios.js [hunk 2]
```diff
@@ -54,10 +54,11 @@ const BpkButtonLink = (props: Props) => {
...
 
-  const themeStyle =
-    theme && themeAttributesSupplied(theme)
-      ? { color: theme.buttonLinkTextColor }
-      : null;
+  const themeAttributes = getThemeAttributes(REQUIRED_THEME_ATTRIBUTES, theme);
+  let themeStyle = null;
+  if (themeAttributes) {
+    themeStyle = { color: themeAttributes.buttonLinkTextColor };
+  }
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-button-link/src/BpkButtonLink.ios.js]: Same as android.

File: native/packages/react-native-bpk-component-button-link/src/utils-test.js [hunk 3]
```diff
@@ -1,56 +0,0 @@
-/*
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-button-link/src/utils-test.js]: 🙌

File: native/packages/react-native-bpk-component-button-link/src/utils.js [hunk 4]
```diff
@@ -1,46 +0,0 @@
-/*
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-button-link/src/utils.js]: 🙌

File: native/packages/react-native-bpk-component-spinner/src/BpkSpinner.js [hunk 5]
```diff
@@ -87,8 +86,8 @@ const BpkSpinner = props => {
 
 const propTypes = {
   small: PropTypes.bool,
-  theme: themePropType,
-  type: PropTypes.oneOf(SPINNER_TYPES),
+  theme: makeThemePropType(REQUIRED_THEME_ATTRIBUTES),
+  type: PropTypes.oneOf(Object.keys(SPINNER_TYPES)),
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-spinner/src/BpkSpinner.js]: Minor: `Object.keys(SPINNER_TYPES)` is done three times - could be one.

File: native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap [hunk 6]
```diff
@@ -64,11 +61,6 @@ exports[`Android BpkSwitch should support theming 1`] = `
   onStartShouldSetResponder={[Function]}
   onTintColor="rgb(230, 228, 235)"
   style={undefined}
-  theme={
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap]: It looks like the theme functionality on the switch has been regressed...

Comment (k0nserv) [reply native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap]: Nope it just used to be spread on the Switch element as part of `rest` before, but now it isn't.

File: native/packages/react-native-bpk-component-button-link/src/BpkButtonLink.android.js (hunk 1 above)
Comment (k0nserv) [reply native/packages/react-native-bpk-component-button-link/src/BpkButtonLink.android.js]: Yeah you're right

File: native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap (hunk 6 above)
Comment (matthewdavidson) [reply native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap]: Ok then in that case we must be missing a test for theming? For instance I can't see any application of a "theme" in these snapshots: https://github.com/Skyscanner/backpack/blob/76b1c7b0a028f312c46b8d15ad57c97cae6a094d/native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap

Comment (k0nserv) [reply native/packages/react-native-bpk-component-switch/src/__snapshots__/BpkSwitch-test.android.js.snap]: The `theme` prop never ends up in the snapshot for any of our themeable components instead other props/styles are derived and passed down. In this case this is captured by the different tint colros

**Pull request merged**